### ISSUE TITLE
HubConnection StartAsync in hosted PWA Blazor Webassembly takes too long in chromium based browsers

### ISSUE THREAD

**Issue by wibram** [dotnet/aspnetcore#47709]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

HubConnection.StartAsync() takes sometimes up to 15 seconds in chromium based browsers, not in Firefox. Tested in Edge 112, Chrome 112 and in Firefox 111.

I am using SignalR websockets for the communication between the blazor.client and blazor.server projects in a hosted Blazor Webassembly PWA. The application is running with several Web APIs in a kubernetes cluster in a cloud system (openstack/euleros).

Usage and StartAsync-Call in Blazor.Client:
```

var hubConnection = new HubConnectionBuilder()
.WithUrl( Navigator.ToAbsoluteUri( "/testhub" ), options =>
{
    options.Transports = HttpTransportType.WebSockets;
    options.SkipNegotiation = true;
})
.WithAutomaticReconnect()
.Build();

hubConnection.On<MyDataModel>( "ReceiveData", ( result ) => HandleData( result ) );

try
{
    // !! sometimes this call lasts up to 15 seconds !!
    await hubConnection.StartAsync();
}
catch (OperationCanceledException) { }
```

SignalR configuration in Program.cs in Blazor.Server:

```
builder.Services.AddSignalR( options =>
{
    options.ClientTimeoutInterval = TimeSpan.FromSeconds( 60 );
    options.HandshakeTimeout = TimeSpan.FromSeconds( 30 );
    options.MaximumParallelInvocationsPerClient = 3;
    options.KeepAliveInterval = TimeSpan.FromSeconds( 5 );
    options.EnableDetailedErrors = true;
} );

// Response compression for SignalR
builder.Services.AddResponseCompression( opts =>
{
    opts.MimeTypes = ResponseCompressionDefaults.MimeTypes.Concat(
        new[] { "application/octet-stream" } );
} );
```

### Expected Behavior

The HubConnection.StartAsync shouldn't last more than a second or should work as fast as in Firefox.

### .NET Version

SDK 7.0.203 Host 7.0.5

### Anything else?

ASP.NET Core 7.0.5
Visual Studio 2022 Community
Running in a kubernetes cluster in a cloud system (openstack/euleros) with several web apis, a second web application and a hashicorp vault, behind a load balancer and ingress nginx controller, ocelot gateway Version 18, Redis Version 6.2 and RabbitMQ Version 3.9.

**Comment by BrennanConroy** [comment]
Turn on logging or get a network trace to see if you can determine what is taking a long time
https://learn.microsoft.com/aspnet/core/signalr/diagnostics?view=aspnetcore-7.0

> Running in a kubernetes cluster in a cloud system (openstack/euleros) with several web apis, a second web application and a hashicorp vault, behind a load balancer and ingress nginx controller, ocelot gateway Version 18, Redis Version 6.2 and RabbitMQ Version 3.9.

Also, remove as many of these as possible to minimize what might be causing the slowdown.

**Comment by wibram** [comment]
I was not able to activate logging. I am getting a PlatformNotSupported Error after starting. In Visual Studio at the creation of the hubconnection at
```
var hubConnection = new HubConnectionBuilder()
.WithUrl(...)
.WithAutomaticReconnect()
.ConfigureLogging( logging => 
{
    logging.AddConsole(); // *
    logging.SetMinimumLevel(LogLevel.Debug);
    logging.AddFilter("Microsoft.AspNetCore.SignalR", LogLevel.Debug);
    logging.AddFilter("Microsoft.AspNetCore.Http.Connections", LogLevel.Debug);
}
.Build();
```

* I had the warning: "Logging this call site is reachable on all platforms. ConsoleLoggerExtensions.AddConsole(ILoggingBuilder) is unsupported on: 'browser'."

**Some more information and tries:**

I am also using a Grpc service to read and write cookies, because in the process of the hub requests/responses through SignalR websockets I can't read/write cookies. So I have Grpc and SignalR websocket communication both with http/2 by default in .NET 7. The chromium browsers have http/2 websockets enabled by default but not in firefox. After enabling that in firefox (config: network.http.spdy.websockets) the Grpc service was not working anymore. After undoing this I removed Grpc from my project and made the calls through web api (controller actions in blazor.server) to read/write cookies but unfortunately the problem with the slow StartAsync call persists. I had one call that lasts 18 seconds.

In each razor page I have several hubconnections. For the razor pages I have created a base class which has a method to create the hub connections:
```
    protected HubConnection CreateHubConnection( string hubPath, bool useAnonymous = false )
    {
        var hubConnection = new HubConnectionBuilder()
            .WithUrl( Navigator.ToAbsoluteUri( hubPath ), options =>
            {
                // using securitytoken
                options.AccessTokenProvider = () => Task.FromResult( ((MyAuthStateProvider)AuthStateProv).Token );
                options.Transports = Microsoft.AspNetCore.Http.Connections.HttpTransportType.WebSockets;
                options.SkipNegotiation = true;
            })
            .Build();

        hubConnection.KeepAliveInterval = TimeSpan.FromSeconds( 5 );

        if ( !useAnonymous)
        {
            // to refresh the securitytoken
            hubConnection.On<ClaimsData>( "ReceiveClaims", ReceiveClaimsAsync );
            // logout when token is missing
            hubConnection.On( "HandleMissingSecurityToken", HandleMissingSecurityToken );
        }

        return hubConnection;
    }
```

I am using Jwt SecurityToken for the authentication in the apis (microservices). So I need to store the token in the cookie and send it with any api call to the blazor.server. The blazor server is using httpclient api calls with the token in the bearer authentication header. If the token is expired I am trying to refresh it with a refreshtoken, which is also stored in a cookie. This is the reason for the ReceiveClaims and HandleMissingSecurityToken extra calls when needed for any hub connection.

In the code section of the razor page which inherits from the class above this is how it looks like:
```
    protected override async Task OnInitializedAsync()
    {
        hubConnectionUser = CreateHubConnection("HUBPath_User");
        hubConnectionRole = CreateHubConnection("HUBPath_Role");
        hubConnectionSite = CreateHubConnection("HUBPath_Site");

        hubConnectionUser.On<UserModel>("ReceiveUser", async (result) => await ReceiveUser(result));
        hubConnectionUser.On<ContactModel>("ReceiveUserContact", (result) => ReceiveUserContact(result));
        hubConnectionRole.On<RoleModel>("ReceiveRole", async (result) => await ReceiveRole(result));
        hubConnectionSite.On<SiteModel>("ReceiveSite", (result) => ReceiveSite(result));

        try
        {
            await hubConnectionUser.StartAsync(); // sometime takes too long
            await hubConnectionRole.StartAsync(); // sometime takes too long
            await hubConnectionSite.StartAsync(); // sometime takes too long
        }
        catch ( OperationCanceledException ) { }
    }
```

It seems that using Grpc and SignalR websockets side by side for the communication between blazor.client and blazor.server is not affecting anything. I don't know if the usage with a base class is a problem to create the hubconnections or using several hubconnections in the same razor page or using several calls for the same hubconnection. I also was not able to debug/log the hubconnection in this environment.

**Comment by BrennanConroy** [comment]
> I was not able to activate logging. I am getting a PlatformNotSupported Error after starting.

Oh right, you're using webassembly:
https://learn.microsoft.com/aspnet/core/blazor/fundamentals/logging?view=aspnetcore-7.0#signalr-client-logging-blazor-webassembly

It's also useful to get the corresponding server logs at the same time you collect the client logs.

> websocket connections over http/1.1 in chromium browsers and also the same way in firefox. I thought that this is default in .NET 7

Browsers don't cache the fact that a server supports http/2 websockets, so if you wait ~30 seconds or so without sending an HTTP/2 request via REST, then the browser will forget that the endpoint supports http/2 websockets and when you connect with a websocket it will use http/1.1

**Comment by wibram** [comment]
But it starts immediately with http/1.1 without waiting. How can I force SignalR Websockets to run with http/2? Or is this not possible because there is also a rest api call which uses http/2 between blazor.client and blazor.server? I changed the Grpc-Call with a Rest API call for reading/writing the cookie because I thought that this causes the problem. I also changed the firefox-config (network.http.spdy.websockets) to use websockets with http/2 but http/1.1 is used anyway. Finally in firefox it runs fine but in all chromium browsers hubconnection.startasync lasts very long.
